### PR TITLE
update System.d.ts

### LINKED ISSUE #113: System API Docs
The [System docs](https://ecsy.io/docs/#/api/classes/syst) appear to be missing a lot of information. Namely `static queries`, `init()`, and `execute(delta, time)`.

How would I go about adding in that information? Is this page auto-generated? Or would I go in and modify it manually?

firefox_2019-10-09_09-20-18

### PR DESCRIPTION
**Pull request by dannyfritz**
Here is a small go at it. It feels off though, I'm not sure if this is really the right place to put these types. The `static queries` also is hard to document.

Closes #113

### CODE REVIEW

File: src/System.d.ts [hunk 1]
```diff
@@ -4,19 +4,34 @@ import {Component, ComponentConstructor} from "./Component";
  * A system that manipulates entities in the world.
  */
 export abstract class System {
+  /**
+   * Defines what Components the System will query for.
+   * This needs to be user defined.
+  */
+  static queries: {
+    [queryName: string]: Component[],
```
Comment (wingyplus) [inline src/System.d.ts]: After I read class System. I think queries might be: 

```
static queries: {
  [queryName: string]: { 
    components: Component[],
    mandatory?: boolean, 
    listen?: "added" | "removed" | "changed"
  }
}
```

For field `components`, I see in example that this field accepted component query but the component query in [this](https://github.com/MozillaReality/ecsy/blob/dev/src/System.d.ts#L29) file, the `Not` selector returns `object` not components. I'm not sure what should we do in this case.

Comment (dannyfritz) [reply src/System.d.ts]: Hrmm, is `listen` also an array of those 3 string consts?

Comment (wingyplus) [reply src/System.d.ts]: @dannyfritz I see System for each valid events that declared in this [line](https://github.com/MozillaReality/ecsy/blob/dev/src/System.js#L55) and get them from `listen`. 

Please suggest me if I'm wrong.

Comment (dannyfritz) [reply src/System.d.ts]: Ah yeah, it appears to be:
```js
listen?: {
  added?: boolean,
  removed?: boolean,
  changed?: boolean,
}
```

Comment (wingyplus) [reply src/System.d.ts]: Ah you're correct. :)

**Comment by wingyplus** [comment]
I see in example that components query accepted `Not` function to do negate operator. And this function return object not class `Component`.  So the type of `queries` object might be: 

```
static queries: {
  [queryName: string]: { 
    components: Component | object[], // because of Not() returns object.
    mandatory?: boolean, 
    listen?: { added?: boolean, removed?: boolean, changed?: boolean }
  }
}
```

**Comment by dannyfritz** [comment]
> I see in example that components query accepted `Not` function to do negate operator. And this function return object not class `Component`. So the type of `queries` object might be:
> 
> ```ts
> static queries: {
>   [queryName: string]: { 
>     components: Component | object[], // because of Not() returns object.
>     mandatory?: boolean, 
>     listen?: { added?: boolean, removed?: boolean, changed?: boolean }
>   }
> }
> ```
> 
> Not sure am I correct? @dannyfritz @brianpeiris

That is probably correct, but in terms of typings seems misleading. Can we create an interface for `Not` to make it more explicit?

File: src/System.d.ts (hunk 1 above)
Comment (dannyfritz) [reply src/System.d.ts]: This isn't valid TypeScript. :cry: 
```ts
  static queries: {
    [queryName: string]: (Component | NotComponent)[],
    listen?: {
      added?: Boolean,
      removed?: Boolean,
      changed?: Boolean | Component[],
    },
  };
```

Comment (dannyfritz) [reply src/System.d.ts]: Also not allowed :frowning_face: 
```ts
static queries: {
    [queryName: Exclude<string, "listen">]: (Component | NotComponent)[],
    listen?: {
      added?: Boolean,
      removed?: Boolean,
      changed?: Boolean | Component[],
    },
  };
```

Comment (dannyfritz) [reply src/System.d.ts]: Oh, this typing is just wrong. XD. Will fix.

Comment (dannyfritz) [reply src/System.d.ts]: got it:

```ts
  static queries: {
    [queryName: string]: {
      components: (Component | NotComponent)[],
      listen?: {
        added?: Boolean,
        removed?: Boolean,
        changed?: Boolean | Component[],
      },
    }
  };
```

**Comment by lambdatastic** [comment]
Doesn't there also need to be a typing for the instance queries in addition to the static queries? I was messing around with the typings a few days ago, and came up with this incomplete interface:

```typescript
interface QueryMap {
  [key: string]: {
    components: Component[]
  }
}

interface QueryResult {
  [key: string]: {
    results: Entity[],
  }
}

/**
 * A system that manipulates entities in the world.
 */
export abstract class System {
  static queries: QueryMap;

  queries: QueryResult;

  // Rest of the system
}
```

**Comment by wingyplus** [comment]
I try convert `factory` example into Typescript and check it with unit test within this project. `System.d.ts` might be:

```typescript
import { Component, ComponentConstructor } from "./Component";
import { Entity } from "ecsy/src/Entity";

/**
 * A system that manipulates entities in the world.
 */
export abstract class System {
  static queries: {
    [queryName: string]: {
      components: (Component | NotComponent)[],
      mandatory?: boolean,
      listen?: {
        added?: boolean,
        removed?: boolean,
        changed?: boolean,
      }
    }
  };

  queries: {
    [queryName: string]: {
      results: Entity[]
    }
  }

  /**
   * Whether the system will execute during the world tick.
   */
  enabled: boolean;
  /**
   * Resume execution of this system.
   */
  play(): void

  /**
   * Stop execution of this system.
   */
  stop(): void
}

export interface SystemConstructor<T extends System> {
  new(...args: any): T;
}

interface NotComponent {
  operator: "not",
  component: Component
}

/**
 * Use the Not class to negate a component query.
 */
export function Not<T>(Component: ComponentConstructor<T>): NotComponent
```

Extract type signature from both `static queries` and `queries` might be a good idea. But I don't have a good name for this. haha :D

**Comment by dannyfritz** [comment]
We're almost there! :joy:

File: src/System.d.ts [hunk 2]
```diff
@@ -1,29 +1,70 @@
...
+import { Entity } from "./Entity";
 
 /**
  * A system that manipulates entities in the world.
  */
 export abstract class System {
+  /**
+   * Defines what Components the System will query for.
+   * This needs to be user defined.
+  */
```
Comment (wingyplus) [inline src/System.d.ts]: inserting space before `*/`.

**Comment by nejcjelovcan** [comment]
Is this planned to be merged?

In addition to the `queries` static and instance variables, I am missing `world` instance viariable on the System. My approach was to declare System as `System<T extends World = World>` and then declare `world:T` instance variable. This allows me to extend Systems that expect a certain world. E.g. a `CachingSystem` expects a `World` that implements `WorldCache` interface.

My branch as an example: https://github.com/nejcjelovcan/ecsy/blob/feature/minimal-typescript-updates/src/System.d.ts (my solution has a similar declaration of `.queries` instance variable as this merge request, but no static `queries` variable).

If theres more work to be done to finally wrap this, I can help...

**Comment by fernandojsg** [comment]
Hi @nejcjelovcan I don't use typescript so I can't really evaluate what is going on here. 
pinging @robertlong or @netpro2k, what do you think about this? can we merge it if it's useful for people and then fixing it with a future PR later on?

**Comment by nickschot** [comment]
We're already using the type definitions from this commit. Big improvement over what was there.

@fernandojsg may I ask why Typescript is not being used to write the lib? It seems like a lot of effort/prone to mistakes to keep the type definitions in sync. As ECSY is already compiled anyway it shouldn't make much difference for users?

**Comment by dannyfritz** [comment]
Doing this PR inspired me to write my own ECS in Typescript that can benefit from generics and type inferences. And it can be done and is very nice to work with it turns out.

File: src/System.d.ts [hunk 3]
```diff
@@ -1,29 +1,70 @@
...
+  static queries: {
+    [queryName: string]: {
+      components: (Component | NotComponent)[],
+      listen?: {
+        added?: Boolean,
+        removed?: Boolean,
+        changed?: Boolean | Component[],
+      },
+    }
+  };
```
Comment (snowfrogdev) [inline src/System.d.ts]: We need to add a `mandatory?: boolean` property. Also according to TS documentation I think we should be using the `boolean` type and not `Boolean`. See https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

I'd be happy to make the fix if the maintainers show signs of wanting to merge this PR in the near future.

Comment (fernandojsg) [reply src/System.d.ts]: @snowfrogdev I'd be happy to get this merge if you are willing to update it. I'm not a typescript dev so I couldn't help on the feedback on the current PR but if you feel the current change is useful, as it's really just changing the .ts definitions, I'm happy to get this merge asap and we could fix the rest of the issues in future PRs.

**Comment by snowfrogdev** [comment]
@dannyfritz Thanks for the update. What about the `mandatory` property. It is not documented at the moment but it is clearly there as we can see in this code example on line 57. https://github.com/MozillaReality/ecsy/blob/dev/examples/canvas/systems.js

I don't really know what it does but I noticed it when copying the canvas example project and TypeScript started complaining about it.

File: src/System.d.ts (hunk 3 above)
Comment (snowfrogdev) [reply src/System.d.ts]: @fernandojsg I work almost exclusively with TypeScript and I'll be happy to help out. It's safe for you to merge this PR in at any time. It is a big improvement over what we have at the moment. Like you said, we can make further improvements in future PRs. Thanks for the quick response by the way. I just started using your library and I was a bit worried you had gone dormant. 😉

**Comment by dannyfritz** [comment]
> @dannyfritz Thanks for the update. What about the `mandatory` property. It is not documented at the moment but it is clearly there as we can see in this code example on line 57. https://github.com/MozillaReality/ecsy/blob/dev/examples/canvas/systems.js
> 
> I don't really know what it does but I noticed it when copying the canvas example project and TypeScript started complaining about it.

I think for the most part, I am done working on this PR. Hopefully someone such as yourself can either base a new PR off of it, or it gets merged and someone such as you can continue the work.

**Comment by snowfrogdev** [comment]
@fernandojsg what is the purpose of the `mandatory` property on the `static queries` object? I can quickly add it to the type definition and documentation before you merge this PR in.

**Comment by fernandojsg** [comment]
@snowfrogdev my fault, I thought `mandatory` was in the docs but I just noticed is not O:)
Basically it means that the `execute` method won't get call if there is a query with `mandatory: true` and the query doesn't have any results so far. 
You can check its tests: https://github.com/MozillaReality/ecsy/blob/dev/test/unit/system.test.js#L709-L772

So do you want me to merge this PR and you could submit a second PR including the `mandatory` ?

**Comment by brianpeiris** [comment]
Just wanted to put in my 2 cents. Even though I wrote the original type script definitions, I'm not particularly well versed in typescript myself (my main goal was to focus on the documentation generation). That said, I'd suggest in future we accept these PRs more readily, since the community is actually keen on using the definitions and is clearly willing to help.
Perhaps @robertlong can participate after the fact and help clean up or improve the community contributions if he's still interested in supporting typescript in ECSY.

**Comment by fernandojsg** [comment]
Yep, I'd be happy if someone else from the community want to review the PR and provide feedback and I'll just merge it right away as I can't really evaluate the PR correctly.

**Comment by brianpeiris** [comment]
Considering @wingyplus has already done a review and @nickschot mentioned that they are already using it, I would say this PR should be merged as is, and future PRs can build on top of it.

**Comment by fernandojsg** [comment]
yep, let's do it and just keep moving it forward, thank you all!

**Pull request merged**